[Issue #3093] Fix T::Range validation on limitless ranges

AlexEvanczuk opened the pull request:
# Summary

Resolves https://github.com/sorbet/sorbet/issues/3093

```ruby
development 30> Range.new(nil, nil).last
RangeError: cannot get the last element of endless range
from (pry):36:in `last'
development 31> Range.new(nil, nil).end
nil
development 32> Range.new(nil, nil).first
nil
development 33> Range.new(nil, nil).begin
nil
```

The docs (https://ruby-doc.org/core-2.5.1/Range.html) state:

> Range#first
> Returns the first object in the range, or an array of the first n elements.
> 
> Range#last
> Returns the last object in the range, or an array of the last n elements.
> 
> Range#begin
> Returns the object that defines the beginning of the range.
> 
> Range#end
> Returns the object that defines the end of the range.

For type validation, we are more interested in the object that *defines* the beginning/ends of the range rather than the literal first/last object.

### Motivation
This allows people to use limitless (no end limit) ranges with Sorbet.

### Test plan
I added tests for these cases.

AlexEvanczuk commented:
@jez Didn't realize there is another place. Could you help me figure out how to run tests locally? I don't see anything in the main readme for how to run sorbet-runtime tests.

AlexEvanczuk commented:
Oh -- and I made a change that should fix that issue you just saw.

jez commented:
You can run the tests with

```
cd gems/sorbet-runtime
rake test
```

jez commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -50,7 +50,7 @@ def valid?(obj)
         return true
       when Range
-        @type.valid?(obj.first) && @type.valid?(obj.last)
+        @type.valid?(obj.begin) && @type.valid?(obj.end)
```
I don't think this is going to work, because if the range is endless, `obj.end` will be `nil` but `nil` will not be a valid `Integer` for e.g. `T::Range[Integer]`

jez replied:
Can you try adding a test like the one you wrote in #3093?

You can look in `method_validation.rb` for how to write tests that test calling real methods with real sigs by giving them real values (rather than by calling into the internals)

AlexEvanczuk replied:
Right -- I'll go with the implementation approach you suggested in the original issue 👌

paracycle commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -124,7 +124,7 @@ def describe_obj(obj)
         T::Hash[inferred_key, inferred_val]
       when Range
-        T::Range[type_from_instances([obj.first, obj.last])]
+        T::Range[type_from_instances([obj.begin, obj.end])]
```
I think this might also be problematic, since `Range.new(1, nil)` would result in being typed as `T::Range[T.nilable(Integer)]` due to the fact that `obj.end` would be `nil`, which is not what you would really expect.

AlexEvanczuk commented:
@jez @paracycle 
I changed the implementation to ensure that the types determined by the runtime match those determined statically:
https://sorbet.run/#T.reveal_type((1...1))%0AT.reveal_type((1...nil))%0A%0AT.reveal_type((nil...1))%0AT.reveal_type((nil...nil))

I added tests to verify this.

Let me know what you think!

paracycle commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -124,7 +128,13 @@ def describe_obj(obj)
-        T::Range[type_from_instances([obj.first, obj.last])]
+        # If the range end is limitless, we cannot call `Range#last`, and the static type checker should consider
+        # it a range of the type of the start of the range.
+        if !obj.begin.nil? && obj.end.nil?
+          T::Range[type_from_instances([obj.begin])]
+        else
+          T::Range[type_from_instances([obj.begin, obj.end])]
+        end
```
How about:
```suggestion
        objs = [obj.begin, obj.end].compact
        objs = T.untyped if objs.empty?
        T::Range[type_from_instances(objs)]
```

paracycle commented on gems/sorbet-runtime/lib/types/types/typed_enumerable.rb:
```diff
@@ -50,7 +50,11 @@ def valid?(obj)
         return true
       when Range
-        @type.valid?(obj.first) && @type.valid?(obj.last)
+        if obj.end.nil?
+          @type.valid?(obj.begin)
+        else
+          @type.valid?(obj.begin) && @type.valid?(obj.end)
+        end
```
Can't we validate the `begin` and `end` separately?
```suggestion
        (obj.begin.nil? || @type.valid?(obj.begin))
          && (obj.end.nil? || @type.valid?(obj.end))
```
This way, any open ends would automatically be valid and if both ends are open, then it would always pass.

AlexEvanczuk commented:
@paracycle Both of those changes appear to fail one of my tests:
When the return type is a non-nilable integer and I return a range with no start and no end, it should pass, but making change 1 causes this failure:
```
  1) Failure:
validation::ranges::return type is non-nilable integer#test_0004_rejects a range with no start and no end [/Users/alex.evanczuk/workspace/sorbet/gems/sorbet-runtime/test/types/method_validation.rb:357]:
Expected /Return value: Expected type T::Range\[Integer\], got T::Range\[NilClass\]/ to match "Return value: Expected type T::Range[Integer], got T::Range[T::Types::Untyped]".

641 runs, 1325 assertions, 1 failures, 0 errors, 1 skips
```

And change two causes this failure:
```
  1) Failure:
validation::ranges::return type is non-nilable integer#test_0004_rejects a range with no start and no end [/Users/alex.evanczuk/workspace/sorbet/gems/sorbet-runtime/test/types/method_validation.rb:352]:
TypeError expected but nothing was raised.

641 runs, 1323 assertions, 1 failures, 0 errors, 1 skips
```

AlexEvanczuk commented:
@jez Happy to make those changes, but currently those suggestions fail the tests -- let me know if you'd like me to remove the failing tests (i.e. not support that behavior).

jez commented:
> @jez Happy to make those changes, but currently those suggestions fail the tests -- let me know if you'd like me to remove the failing tests (i.e. not support that behavior).

Sorry I left my review before seeing your comment.

```
irb(main):004:0> x = (nil..nil)
=> nil..
irb(main):005:0> x
=> nil..
irb(main):006:0> x.begin
=> nil
irb(main):007:0> x.end
=> nil
irb(main):008:0> x.first
=> nil
irb(main):009:0> x.last
Traceback (most recent call last):
        5: from /Users/jez/.rbenv/versions/2.6/bin/irb:23:in `<main>'
        4: from /Users/jez/.rbenv/versions/2.6/bin/irb:23:in `load'
        3: from /Users/jez/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):9
        1: from (irb):9:in `last'
RangeError (cannot get the last element of endless range)
```

^ this tells me that the Ruby VM has a separate notion of "endless" that's not just `.begin` / `.end` is `nil`"

That probably means there's a way to check if a range is endless or not, and we should use that instead of checking `.begin` / `.end`

Do you mind tracking that down?

AlexEvanczuk commented:
@jez 
Looks like there is an existing ruby feature request to give such a method: https://bugs.ruby-lang.org/issues/15864

jez commented:
It looks like sadly the behavior has changed on my example from Ruby 2.6 to Ruby 2.7.

AlexEvanczuk commented:
What behavior differences are you seeing Jez? What's the process for accounting for different behaviors in different versions of Ruby in sorbet-runtime? Happy to incorporate some adapter for different ruby versions.

jez commented:
```
irb(main):004:0> (nil..nil)
=> nil..nil
irb(main):005:0> x = (nil..nil)
irb(main):006:0> x
=> nil..nil
irb(main):007:0> x.begin
=> nil
irb(main):008:0> x.end
=> nil
irb(main):009:0> x.first
Traceback (most recent call last):
        5: from /Users/jez/.rbenv/versions/2.7.1/bin/irb:23:in `<main>'
        4: from /Users/jez/.rbenv/versions/2.7.1/bin/irb:23:in `load'
        3: from /Users/jez/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        2: from (irb):9
        1: from (irb):9:in `first'
RangeError (cannot get the first element of beginless range)
irb(main):010:0> x.last
Traceback (most recent call last):
        6: from /Users/jez/.rbenv/versions/2.7.1/bin/irb:23:in `<main>'
        5: from /Users/jez/.rbenv/versions/2.7.1/bin/irb:23:in `load'
        4: from /Users/jez/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        3: from (irb):9
        2: from (irb):10:in `rescue in irb_binding'
        1: from (irb):10:in `last'
RangeError (cannot get the last element of endless range)
```

There isn't really a process.

At this point, I'd probably just rescue the `RangeError` and call it a day.

AlexEvanczuk commented:
Could you do just run this for me and tell me your output:

```ruby
range_starts = [1, nil]
range_ends = [10, nil]

range_starts.map do |range_start|
  range_ends.map do |range_end|
    puts "======================"
    puts "======================"
    puts "Range.new(#{range_start}, #{range_end})"
    range = nil
    begin
      range = Range.new(range_start, range_end)
    rescue => ex
      puts "Failed to initialize this range with error: #{ex.class}: #{ex.message}"
      next
    end

    puts "range.first"
    puts (range.first.inspect rescue "failed to call first")
    puts "range.last"
    puts (range.last.inspect rescue "failed to call last")
    puts "range.begin"
    puts range.begin.inspect
    puts "range.end"
    puts range.end.inspect
  end
end;
```
```
======================
======================
Range.new(1, 10)
range.first
1
range.last
10
range.begin
1
range.end
10
======================
======================
Range.new(1, )
range.first
1
range.last
failed to call last
range.begin
1
range.end
nil
======================
======================
Range.new(, 10)
Failed to initialize this range with error: ArgumentError: bad value for range
======================
======================
Range.new(, )
range.first
nil
range.last
failed to call last
range.begin
nil
range.end
nil
```

paracycle commented:
Are we interested in the behaviour of `last` and `first` at all? It looks like `begin` and `end` do signal the limits properly in all versions. Am I missing something?

AlexEvanczuk commented:
> Are we interested in the behaviour of `last` and `first` at all? It looks like `begin` and `end` do signal the limits properly in all versions. Am I missing something?

No, I don't think you're missing something. I'm not sure the behavior of `first` or `last` is relevant to determining the proper types, as long as we are switching properly on `begin` and `end`.

paracycle commented:
Ok. I think I know why some tests are failing with my proposed changes. Here are a few observations:

1. We really should not be testing for Ruby 2.6 failures on beginless ranges. I think we should only run those tests if Ruby version is >= 2.7.0, since on those versions, the tests run properly. I know that is not the CI setup of Sorbet at this point, so maybe we don't even need the "beginless" range tests until Sorbet supports and is tested on 2.7.
2. Many of the return types in the tests are `T::Range[T.nilable(Integer)]` but those should all be `T::Range[Integer]` since, even though the begin or the end is not defined, those are still `Integer` ranges without any `nil`s.

jez commented:
> I know that is not the CI setup of Sorbet at this point, so maybe we don't even need the "beginless" range tests until Sorbet supports and is tested on 2.7.

If you write a test that is expected to do different things on different ruby versions, i'd appreciate if you guarded it by a check of the Ruby version, so that we don't have to track down the intended meaning of the test when doing a ruby upgrade. Here's an example test:

https://github.com/sorbet/sorbet/pull/3052/files#diff-c68234159e068097095fb22edbed4977R479

paracycle commented:
> > I know that is not the CI setup of Sorbet at this point, so maybe we don't even need the "beginless" range tests until Sorbet supports and is tested on 2.7.
> 
> If you write a test that is expected to do different things on different ruby versions, i'd appreciate if you guarded it by a check of the Ruby version, so that we don't have to track down the intended meaning of the test when doing a ruby upgrade. Here's an example test:
> 
> https://github.com/sorbet/sorbet/pull/3052/files#diff-c68234159e068097095fb22edbed4977R479

Ah right, I forgot that we already had those. In that case @AlexEvanczuk, I recommend that the tests that involve begin-less ranges (`(nil..10)` or `(nil..nil)`) are marked as passing `if RUBY_VERSION >= "2.7"` tests.

AlexEvanczuk commented:
I think I'm a still bit confused, looking for a bit more clarification before I proceed.

It looks like Ruby 2.6 and Ruby 2.7 behave slightly differently.
Ruby 2.6 allows you to get the `first` but not the `last` item of a range if the first/last elements are `nil`. Ruby 2.7 makes this more consistent by having `first` fail if the first element is `nil` as well. This makes sense from the ruby perspective, but I don't see the connection to how Sorbet thinks about these types.

Beginless ranges appear to only work on both Ruby 2.6 and 2.7 by erroring if the first argument is `nil` (so neither versions need this test), if the second arg is `nil`, then its a `T::Range[NilClass]`. Regardless of whether we're in ruby 2.6 or 2.7, shouldn't the types that sorbet-runtime infer be the same? Since we're not using `last` or `first` this difference shouldn't affect Ruby 2.7.

Let me know what I'm missing!

AlexEvanczuk commented:
Thanks for clarifying @paracycle . That 2.7 does not error on a begin-less range is what I was missing here (hence the little Ruby script to see the different behaviors above).

The fact that Ruby 2.7 does not error on begin-less ranges makes me happy though, since the fact that Ruby 2.6 only supports limitless at the end but not the beginning didn't make much sense to me.

I agree that if we have `(nil...nil)`, it doesn't mean its a range of nils, but a boundless range of potentially anything, so calling it `T::Range[T.untyped]` makes a lot of sense. To put it another way, the code should not use the existence of a `nil` to try to infer anything about the type (i.e. boundlessness offers no type information).

These are the changes I will make:
1) I will add a beginless range test, but if we're on V2.7, the test will deliberately error (asking the team to implement that test as part of adding multi-ruby test suites). When on 2.6, the test will assert an argument error.
2) I will incorporate your suggested changes above to stop using `nil` to attempt to infer type information.

I think we're aligned on this now! Let me know if you feel otherwise!

paracycle commented:
Thanks. We are agreed except for one small thing:

> 1. I will add a beginless range test, but if we're on V2.7, the test will deliberately error (asking the team to implement that test as part of adding multi-ruby test suites). When on 2.6, the test will assert an argument error.

I don't think we need to check for test failures on 2.6 at all, since that would be testing the failure of Ruby and not Sorbet. After all, the `ArgumentError` is from Ruby and would be expected. It wouldn't be testing any of the Sorbet behaviour.

Let's just create tests that are only run in 2.7 or above Ruby versions, and that test begin-less range cases successfully.

AlexEvanczuk commented:
One more thing @paracycle and @jez, before I proceed, do we intend for sorbet-runtime to be consistent with sorbet static?

If so, sorbet static currently contradicts the conclusions we've made together:
https://sorbet.run/#T.reveal_type((1...1))%0AT.reveal_type((1...nil))%0AT.reveal_type((nil...1))%0AT.reveal_type((nil...nil))%0AT.reveal_type((1...1).begin)%0AT.reveal_type((1...1).end)%0AT.reveal_type((1...nil).begin)%0AT.reveal_type((1...nil).end)%0AT.reveal_type((nil...1).begin)%0AT.reveal_type((nil...1).end)%0AT.reveal_type((nil...nil).begin)%0AT.reveal_type((nil...nil).end)

That is, it considers a range of `(nil...nil)` to be a range of `NilClass`. Similarly, some ranges that have `nil` are `T::Range[T.nilable(Integer)]`. `begin` and `end` carry forth these implications.

I agree with @paracycle from the perspective of fundamental correctness (an unbounded range is not a range of nils, its a range of unknown values), but that contradicts the static type system which could be confusing.

I'll defer to whatever decision the two of you want to make.

My preference is to have sorbet-runtime be consistent with the messaging of sorbet static, to avoid developer confusion and frustration (but my higher preference is change sorbet static to be more "correct" given our new interpretations).

paracycle commented:
@AlexEvanczuk I think the problem there is that Sorbet static does not support any of the new 2.7 features that need the parser to be updated. Beginless ranges are one of those, since they also involve the new `(..10)` syntax, so need to be explicitly added to Sorbet.

We, at Shopify, are in the process of looking at bringing Sorbet static up to date with 2.7 syntax and begin-less ranges will be handled (or would be possible to handle them) once that is done.

But, you are correct that the inferences made by the static checker are not correct from how we've analyzed the situation so far. Not sure what @jez thinks, but I think that needs to be tracked as part of another ticket.

jez commented:
See https://github.com/sorbet/sorbet/issues/972 for a long discussion about what the static behavior might be.

paracycle commented:
> See #972 for a long discussion about what the static behavior might be.

Indeed, the static behaviour is currently correct because it is fully correct on the Ruby 2.6 feature subset. Where the static behaviour starts failing is in the new 2.7 behaviour (note that only the ranges that don't have a `begin` have wrong results in @AlexEvanczuk's link), and that is to be expected for now.

I presume that once we update to 2.7 syntax, begin-less ranges will automatically correct themselves (maybe except the unbounded case).

AlexEvanczuk commented:
"fully correct on Ruby 2.6 feature subset" -> this is a great way to put it, thanks for clarifying. I've [created this `sorbet.run` to capture that point in concrete detail](https://sorbet.run/#T.reveal_type((1...1))%0AT.reveal_type((1...nil))%0A%23%20T.reveal_type((nil...1))%20%23%20only%20on%20ruby%202.7%0AT.reveal_type((nil...nil))%20%23%20this%20is%20the%20only%20one%20that%20has%20the%20%22wrong%22%20static%20types%20on%20ruby%202.6%0AT.reveal_type((1...1).begin)%0AT.reveal_type((1...1).end)%0AT.reveal_type((1...nil).begin)%0AT.reveal_type((1...nil).end)%0A%23%20T.reveal_type((nil...1).begin)%20%23%20only%20on%20ruby%202.7%0A%23%20T.reveal_type((nil...1).end)%20%23%20only%20on%20ruby%202.7%0AT.reveal_type((nil...nil).begin)%0AT.reveal_type((nil...nil).end)). It reveals one inconsistency, which is that `(nil..nil)` is supported on Ruby 2.6, and is considered `T::Range[NilClass]` and not `T::Range[T.untyped]`.

I'll make a change so the Ruby 2.6 unsupported features are not tested (since we don't need to test core ruby). Ruby 2.7 only features will be tested with a guard clause, and those tests will test what we hope to be the correct functionality for beginless ranges.

Thanks for shepherding me through this change! I'll incorporate the latest sometime later today.

AlexEvanczuk commented:
@paracycle and @jez I've made the discussed changes! Let me know what you think.

paracycle commented on gems/sorbet-runtime/test/types/method_validation.rb:
```diff
@@ -306,6 +306,158 @@ def @mod.foo
+          end
+        end
+
+        describe 'return type is nilable integer' do
```
I am not sure if there is any extra information in these tests. Now that we know `T::Range[Integer]` is an allowed return type for these ranges, these tests only show that Sorbet is able to properly treat `T::Range[Integer]`s as `T::Range[T.nilable(Integer)]`s. 

So these tests end up testing that covariance is correctly applied to the generic return type by Sorbet, which is not very relevant to the current test aim. I say we drop these tests.

AlexEvanczuk replied:
Good point!  Dropped.

paracycle commented on gems/sorbet-runtime/test/types/method_validation.rb:
```diff
@@ -306,6 +306,158 @@ def @mod.foo
+          end
+        end
+
+        describe 'return type is a range of nils' do
```
Similar to my comment above, these tests are only testing that `T::Range[Integer]` cannot be cast to `T::Range[NilClass]` (or that `T::Range[T.untyped]` can be cast to `T::Range[NilClass]`, for the final case) and don't convey more information about how Range values are being interpreted. I say we drop these tests.

AlexEvanczuk replied:
Yep makes sense!

AlexEvanczuk commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -369,13 +369,37 @@ def subtype_of?(other)
+      end
+
+      it 'works if the range has no start or end' do
+        type = T::Range[NilClass]
```
yep!

AlexEvanczuk commented on gems/sorbet-runtime/test/types/types.rb:
```diff
@@ -369,13 +369,37 @@ def subtype_of?(other)
+        assert_nil(msg)
+      end
+
+      it 'works if the range has no start or end' do
```
Actually, beginless AND endless ranges are supported in Ruby 2.6 -- it's beginless only ranges that are not supported.

paracycle replied:
Ah cool, even better.

AlexEvanczuk commented:
Thanks for your feedback @paracycle . I've incorporated your suggested changes.

AlexEvanczuk commented:
@jez Could you unblock tests and let me know if you have any other requested changes before merging?

jez commented:
@AlexEvanczuk sorry for not having time for this so far; i've been on call at work this week and it's been unusually busy. i'll need a bit more time before i can get back to you on this.

AlexEvanczuk commented:
No worries @jez thanks for the info! The main reason I want this is because we want to create a `DateRange` gem that inherits from `Range`, but we cannot until sorbet supports the boundless ranges. For now we'll just not inherit from `Range`!

Looking forward to merging when you're not so swamped.

jez commented:
I am going to test this on Stripe's codebase just to be extra careful, but other than that this looks good to me, and I'll merge when that finishes.

jez commented:
@AlexEvanczuk when I tested this on Stripe's codebase it introduced hundreds of runtime exceptions. I'm not sure what the problem is, but I'm looking into it. I'm going to dismiss my approval right now because i think there might be something wrong with this change.

AlexEvanczuk commented:
Thanks jez for looking through and running the smoke test -- glad everything was green once we merged master!
